protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,720 @@
+                let connection = super::RelayedConnection::new_inbound(inbound_circuit, tx);
+
+                pending_msgs.push_back(transport::ToListenerMsg::IncomingRelayedConnection {
+                    stream: connection,
+                    src_peer_id,
+                    relay_peer_id: self.remote_peer_id,
+                    relay_addr: self.remote_addr.clone(),
+                });
+
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::InboundCircuitEstablished {
+                        src_peer_id: self.remote_peer_id,
+                        limit,
+                    },
+                ));
+            }
+            Reservation::None => {
+                let src_peer_id = inbound_circuit.src_peer_id();
+                self.circuit_deny_futs.push(
+                    inbound_circuit
+                        .deny(Status::NoReservation)
+                        .map(move |result| (src_peer_id, result))
+                        .boxed(),
+                )
+            }
+        }
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        output: <Self::OutboundProtocol as upgrade::OutboundUpgrade<NegotiatedSubstream>>::Output,
+        info: Self::OutboundOpenInfo,
+    ) {
+        match (output, info) {
+            (
+                outbound_hop::Output::Reservation {
+                    renewal_timeout,
+                    addrs,
+                    limit,
+                },
+                OutboundOpenInfo::Reserve { to_listener },
+            ) => {
+                let event = self.reservation.accepted(
+                    renewal_timeout,
+                    addrs,
+                    to_listener,
+                    self.local_peer_id,
+                    limit,
+                );
+
+                self.queued_events
+                    .push_back(ProtocolsHandlerEvent::Custom(event));
+            }
+            (
+                outbound_hop::Output::Circuit {
+                    substream,
+                    read_buffer,
+                    limit,
+                },
+                OutboundOpenInfo::Connect { send_back },
+            ) => {
+                let (tx, rx) = oneshot::channel();
+                match send_back.send(Ok(super::RelayedConnection::new_outbound(
+                    substream,
+                    read_buffer,
+                    tx,
+                ))) {
+                    Ok(()) => self.alive_lend_out_substreams.push(rx),
+                    Err(_) => debug!(
+                        "Oneshot to `RelayedDial` future dropped. \
+                         Dropping established relayed connection to {:?}.",
+                        self.remote_peer_id,
+                    ),
+                }
+
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::OutboundCircuitEstablished { limit },
```

Comment:
Do you still want to push this event if the `send_back.send` failed?

Reply:
Good point. Fixed in b1cb6848f2c6c9bbfd699f18649737e28767ada3.